File: client/src-tauri/src/commands/contracts/dictation.rs
use serde::{Deserialize, Serialize};

pub const SPEECH_DICTATION_START_COMMAND: &str = "speech_dictation_start";
pub const SPEECH_DICTATION_STOP_COMMAND: &str = "speech_dictation_stop";
pub const SPEECH_DICTATION_CANCEL_COMMAND: &str = "speech_dictation_cancel";
pub const SPEECH_DICTATION_SETTINGS_COMMAND: &str = "speech_dictation_settings";
pub const SPEECH_DICTATION_UPDATE_SETTINGS_COMMAND: &str = "speech_dictation_update_settings";

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum DictationPlatformDto {
    Macos,
    Windows,
    Linux,
    Unsupported,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum DictationEngineDto {
    Modern,
    Legacy,
    WindowsSdk,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum DictationEnginePreferenceDto {
    Automatic,
    Modern,
    Legacy,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum DictationPrivacyModeDto {
    OnDevicePreferred,
    OnDeviceRequired,
    AllowNetwork,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum DictationPermissionStateDto {
    Authorized,
    Denied,
    Restricted,
    NotDetermined,
    Unsupported,
    Unknown,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum DictationStopReasonDto {
    User,
    Cancelled,
    Error,
    ChannelClosed,
    AppClosing,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum DictationModernAssetStatusDto {
    Installed,
    NotInstalled,
    Unavailable,
    UnsupportedLocale,
    Unknown,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DictationEngineStatusDto {
    pub available: bool,
    pub compiled: bool,
    pub runtime_supported: bool,
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DictationModernAssetsDto {
    pub status: DictationModernAssetStatusDto,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub locale: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ActiveDictationSessionDto {
    pub session_id: String,
    pub engine: DictationEngineDto,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DictationStatusDto {
    pub platform: DictationPlatformDto,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub os_version: Option<String>,
    pub default_locale: Option<String>,
    #[serde(default)]
    pub supported_locales: Vec<String>,
    pub modern: DictationEngineStatusDto,
    pub legacy: DictationEngineStatusDto,
    #[serde(default)]
    pub windows_sdk: DictationEngineStatusDto,
    pub modern_assets: DictationModernAssetsDto,
    pub microphone_permission: DictationPermissionStateDto,
    pub speech_permission: DictationPermissionStateDto,
    pub active_session: Option<ActiveDictationSessionDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DictationSettingsDto {
    pub engine_preference: DictationEnginePreferenceDto,
    pub privacy_mode: DictationPrivacyModeDto,
    pub locale: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub updated_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct UpsertDictationSettingsRequestDto {
    pub engine_preference: DictationEnginePreferenceDto,
    pub privacy_mode: DictationPrivacyModeDto,
    pub locale: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DictationStartRequestDto {
    pub locale: Option<String>,
    pub engine_preference: Option<DictationEnginePreferenceDto>,
    pub privacy_mode: Option<DictationPrivacyModeDto>,
    #[serde(default)]
    pub contextual_phrases: Vec<String>,
    #[serde(default)]
    pub channel: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DictationStartResponseDto {
    pub session_id: String,
    pub engine: DictationEngineDto,
    pub locale: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum DictationEventDto {
    Permission {
        microphone: DictationPermissionStateDto,
        speech: DictationPermissionStateDto,
    },
    Started {
        #[serde(rename = "sessionId")]
        session_id: String,
        engine: DictationEngineDto,
        locale: String,
    },
    AssetInstalling {
        progress: Option<f32>,
    },
    Partial {
        #[serde(rename = "sessionId")]
        session_id: String,
        text: String,
        sequence: u64,
    },
    AudioLevel {
        #[serde(rename = "sessionId")]
        session_id: String,
        level: f32,
        sequence: u64,
    },
    Final {
        #[serde(rename = "sessionId")]
        session_id: String,
        text: String,
        sequence: u64,
    },
    Stopped {
        #[serde(rename = "sessionId")]
        session_id: String,
        reason: DictationStopReasonDto,
    },
    Error {
        #[serde(rename = "sessionId")]
        session_id: Option<String>,
        code: String,
        message: String,
        retryable: bool,
    },
}

File: client/src/lib/xero-model/dictation.ts
import { z } from 'zod'

import { normalizeOptionalText } from './shared'

export const dictationPlatformSchema = z.enum(['macos', 'unsupported'])
export const dictationEngineSchema = z.enum(['modern', 'legacy'])
export const dictationEnginePreferenceSchema = z.enum(['automatic', 'modern', 'legacy'])
export const dictationPrivacyModeSchema = z.enum([
  'on_device_preferred',
  'on_device_required',
  'allow_network',
])
export const dictationPermissionStateSchema = z.enum([
  'authorized',
  'denied',
  'restricted',
  'not_determined',
  'unsupported',
  'unknown',
])
export const dictationStopReasonSchema = z.enum([
  'user',
  'cancelled',
  'error',
  'channel_closed',
  'app_closing',
])
export const dictationModernAssetStatusSchema = z.enum([
  'installed',
  'not_installed',
  'unavailable',
  'unsupported_locale',
  'unknown',
])

const normalizedOptionalTextSchema = z
  .string()
  .nullable()
  .optional()
  .transform((value) => normalizeOptionalText(value))

export const dictationEngineStatusSchema = z
  .object({
    available: z.boolean(),
    compiled: z.boolean(),
    runtimeSupported: z.boolean(),
    reason: normalizedOptionalTextSchema,
  })
  .strict()

export const dictationModernAssetsSchema = z
  .object({
    status: dictationModernAssetStatusSchema,
    locale: normalizedOptionalTextSchema,
    reason: normalizedOptionalTextSchema,
  })
  .strict()

export const activeDictationSessionSchema = z
  .object({
    sessionId: z.string().trim().min(1),
    engine: dictationEngineSchema,
  })
  .strict()

export const dictationStatusSchema = z
  .object({
    platform: dictationPlatformSchema,
    osVersion: normalizedOptionalTextSchema,
    defaultLocale: normalizedOptionalTextSchema,
    supportedLocales: z.array(z.string().trim().min(1)).optional().default([]),
    modern: dictationEngineStatusSchema,
    legacy: dictationEngineStatusSchema,
    modernAssets: dictationModernAssetsSchema
      .optional()
      .default({ status: 'unknown', locale: null, reason: null }),
    microphonePermission: dictationPermissionStateSchema,
    speechPermission: dictationPermissionStateSchema,
    activeSession: activeDictationSessionSchema.nullable().optional().transform((value) => value ?? null),
  })
  .strict()

export const dictationSettingsSchema = z
  .object({
    enginePreference: dictationEnginePreferenceSchema,
    privacyMode: dictationPrivacyModeSchema,
    locale: normalizedOptionalTextSchema,
    updatedAt: normalizedOptionalTextSchema,
  })
  .strict()

export const upsertDictationSettingsRequestSchema = z
  .object({
    enginePreference: dictationEnginePreferenceSchema,
    privacyMode: dictationPrivacyModeSchema,
    locale: normalizedOptionalTextSchema,
  })
  .strict()

export const dictationStartRequestSchema = z
  .object({
    locale: normalizedOptionalTextSchema,
    enginePreference: dictationEnginePreferenceSchema.nullable().optional().transform((value) => value ?? 'automatic'),
    privacyMode: dictationPrivacyModeSchema
      .nullable()
      .optional()
      .transform((value) => value ?? 'on_device_preferred'),
    contextualPhrases: z.array(z.string().trim().min(1)).optional().default([]),
  })
  .strict()

export const dictationStartResponseSchema = z
  .object({
    sessionId: z.string().trim().min(1),
    engine: dictationEngineSchema,
    locale: z.string().trim().min(1),
  })
  .strict()

const dictationSequenceSchema = z.number().int().nonnegative()

export const dictationEventSchema = z.discriminatedUnion('kind', [
  z
    .object({
      kind: z.literal('permission'),
      microphone: dictationPermissionStateSchema,
      speech: dictationPermissionStateSchema,
    })
    .strict(),
  z
    .object({
      kind: z.literal('started'),
      sessionId: z.string().trim().min(1),
      engine: dictationEngineSchema,
      locale: z.string().trim().min(1),
    })
    .strict(),
  z
    .object({
      kind: z.literal('asset_installing'),
      progress: z.number().min(0).max(1).nullable(),
    })
    .strict(),
  z
    .object({
      kind: z.literal('partial'),
      sessionId: z.string().trim().min(1),
      text: z.string(),
      sequence: dictationSequenceSchema,
    })
    .strict(),
  z
    .object({
      kind: z.literal('final'),
      sessionId: z.string().trim().min(1),
      text: z.string(),
      sequence: dictationSequenceSchema,
    })
    .strict(),
  z
    .object({
      kind: z.literal('stopped'),
      sessionId: z.string().trim().min(1),
      reason: dictationStopReasonSchema,
    })
    .strict(),
  z
    .object({
      kind: z.literal('error'),
      sessionId: z.string().trim().min(1).nullable(),
      code: z.string().trim().min(1),
      message: z.string().trim().min(1),
      retryable: z.boolean(),
    })
    .strict(),
])

export type DictationPlatformDto = z.infer<typeof dictationPlatformSchema>
export type DictationEngineDto = z.infer<typeof dictationEngineSchema>
export type DictationEnginePreferenceDto = z.infer<typeof dictationEnginePreferenceSchema>
export type DictationPrivacyModeDto = z.infer<typeof dictationPrivacyModeSchema>
export type DictationPermissionStateDto = z.infer<typeof dictationPermissionStateSchema>
export type DictationStopReasonDto = z.infer<typeof dictationStopReasonSchema>
export type DictationModernAssetStatusDto = z.infer<typeof dictationModernAssetStatusSchema>
export type DictationEngineStatusDto = z.infer<typeof dictationEngineStatusSchema>
export type DictationModernAssetsDto = z.infer<typeof dictationModernAssetsSchema>
export type ActiveDictationSessionDto = z.infer<typeof activeDictationSessionSchema>
export type DictationStatusDto = z.infer<typeof dictationStatusSchema>
export type DictationSettingsDto = z.infer<typeof dictationSettingsSchema>
export type UpsertDictationSettingsRequestDto = z.infer<typeof upsertDictationSettingsRequestSchema>
export type DictationStartRequestInputDto = z.input<typeof dictationStartRequestSchema>
export type DictationStartRequestDto = z.infer<typeof dictationStartRequestSchema>
export type DictationStartResponseDto = z.infer<typeof dictationStartResponseSchema>
export type DictationEventDto = z.infer<typeof dictationEventSchema>
